Tr compiler v2

SarcasticNastik opened the pull request:
This PR builds on top of #291. This aims to introduce an efficient version of the tapscript compiler by using a few heuristics to optimize over the expected average total cost for the TapTree.

## Strategy implemented

- While merging TapTrees `A` and `B`, check whether the compilation of `Policy::Or(policy(A), policy(B))` is more efficient than a TapTree with the respective children `A` and `B`. 

**Note**: This doesn't include the `thresh(k, ...n..)` enumeration strategy. Planning on working on it separately.

sanket1729 reviewed (commented):
Some partial review of the first few commits

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -180,10 +183,31 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
+                Self::taptree_cost(l, ms_cache, policy_cache, depth + 1) + Self::taptree_cost(r, ms_cache, policy_cache, depth + 1)
+            }
+            TapTree::Leaf(ref ms) => {
+                let prob = match ms_cache.get(ms) {
```
Right, missed that! Addressed in 4e2f2fa

SarcasticNastik commented on src/policy/compiler.rs:
```diff
@@ -1117,6 +1117,24 @@ pub fn best_compilation<Pk: MiniscriptKey, Ctx: ScriptContext>(
 }
 
+/// Obtain the best compilation of for p=1.0 and q=0
+pub fn tr_best_compilation<Pk: MiniscriptKey, Ctx: ScriptContext>(
```
Right!
The function is not `tr` specific, but it's required to obtain the miniscript compilation of the policy along with the satisfaction cost for the policy. Renamed it to `best_compilation_sat` for now.

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -255,7 +362,11 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
+    pub fn compile_tr(
+        &self,
+        unspendable_key: Option<Pk>,
+        eff: bool,
```
Addressed in 219119d

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -169,6 +174,88 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
 
+    /// Taproot Tree cost (with `branch_prob`)
+    #[cfg(feature = "compiler")]
+    fn tr_node_cost(ms: &Miniscript<Pk, Tap>, prob: f64, cost: &f64) -> OrdF64 {
```
Addressed in 219119d

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -33,12 +33,17 @@ use {
 
+// Change to Arc<TapTree<Pk>>
+#[cfg(feature = "compiler")]
+type PolicyTapCache<Pk> = BTreeMap<TapTree<Pk>, (Policy<Pk>, f64)>;
```
Addressed in 219119d

SarcasticNastik commented on src/lib.rs:
```diff
@@ -94,7 +94,7 @@
 #![deny(non_camel_case_types)]
 #![deny(non_snake_case)]
 #![deny(unused_mut)]
-#![deny(dead_code)]
```
Addressed in 219119d

SarcasticNastik commented on src/policy/concrete.rs:
```diff
@@ -256,15 +256,24 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
-            },
-        )?;
-        Ok(tree)
+        self.is_valid()?; // Check for validity
```
Update: Following discussions in issue #338, we currently disallow repeated keys in taproot policy.

sanket1729 replied:
My thinking is as follows: we should allow repeated keys in the policy. While doing individual policy compilations to miniscript, we should check this. Not while compiling the entire tr policy. I want to make sure that we support policy compilations that allow same keys in different tree leaves.


As for this PR, this need not concern us.

SarcasticNastik commented:
@sanket1729 This PR is ready-for-review given #291 has been merged now.

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -41,6 +42,14 @@ use crate::miniscript::limits::{LOCKTIME_THRESHOLD, SEQUENCE_LOCKTIME_TYPE_FLAG}
+
+/// [`Miniscript`] -> leaf probability in policy cache
+#[cfg(feature = "compiler")]
+type MsTapCache<Pk> = BTreeMap<TapTree<Pk>, f64>;
```
What happens when there is a collision? Making sure that we are dealing with those cases in a sane manner.

SarcasticNastik replied:
Sure I'll update this! This was done considering that there were no duplicate keys in the initial policy.

sanket1729 replied:
There are no duplicates, but still, there can be collisions in the compilation.

SarcasticNastik replied:
Right, that makes sense!

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -805,6 +870,36 @@ where
+    tr: &TapTree<Pk>,
+    ms_cache: &MsTapCache<Pk>,
+    policy_cache: &PolicyTapCache<Pk>,
+    depth: u32,
```
Why is there a depth parameter here? calculating the cost of the tree at a depth? But then what does average mean?

sanket1729 replied:
Looks like this is always called with value 0. We should remove it. Perhaps you need a helper around it.

SarcasticNastik replied:
Addressed in d384b63

sanket1729 commented on examples/taproot.rs:
```diff
@@ -0,0 +1,44 @@
+
+fn main() {
+    let policies_str = [
+        "or(1@pk(A),1@pk(B))",
```
We should make this example as a tutorial. 

- Take a example policy,
 - explain what does
 - explain the compiled result. 
 
 Do this a couple of times for interesting results. Using unspendable internal key, useful internal key being inferred etc..

sanket1729 reviewed (commented):
approach and algorithm ACK. some code improvments to be done

SarcasticNastik commented:
### Changes in the force-push
1. Addressing error lifting issues.
2. Update taproot example as suggested.
3. Handling collisions in cache and added comments for assumptions.

SarcasticNastik commented:
### Changes in the force-push
1. Addressing error lifting issues.
2. Update taproot example as suggested.
3. Handling collisions in cache and added comments for assumptions.

SarcasticNastik commented:
### Changes in the force-push
1. Addressing error lifting issues.
2. Update taproot example as suggested.
3. Handling collisions in cache and added comments for assumptions.

sanket1729 commented:
@SarcasticNastik, we recently merged support for no-std. This would require another rebase

SarcasticNastik commented:
@sanket1729 Done!

sanket1729 commented:
Hey @SarcasticNastik, looks are still some issues. Let me know if you are having difficulty in dealing with CI with no-std. The general thumb rules are 1) use core instead of std, 2) Common things should be available via `use prelude::*`

SarcasticNastik commented:
@sanket1729 Finally done :p! CI passes for no-std support.

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -857,3 +972,122 @@ fn with_huffman_tree<Pk: MiniscriptKey>(
+        // Obtain compilation for the parent policy
+        let (parent_compilation, parent_sat_cost) =
+            compiler::best_compilation_sat::<Pk, Tap>(&parent_policy)?;
+        parent_compilation.sanity_check()?;
```
Here if the parent compilation fails the sanity check, we should ignore it.

SarcasticNastik replied:
Done!

sanket1729 commented on src/policy/mod.rs:
```diff
@@ -492,5 +494,32 @@ mod tests {
+                .compile_tr(Some(unspendable_key.clone()))
+                .unwrap();
+            let opt_expected_desc = Descriptor::Tr(
+                Tr::<String>::from_str("tr(A,{{pk(D),pk(C)},{pk(B),and_v(v:pk(E),pk(F))}})")
```
Did we succeed in finding any policy where these differs?

SarcasticNastik replied:
We did find policies where the TapTree leaf nodes re-arranged, but none yet where we noticed policy-merging. One of those were:
Policy: `thresh(1,or(1@pk(A),1@pk(B)),or(1@pk(C),1@or(1@and(pk(E),pk(F)),1@pk(D))))`
Private compilation: `tr(A,{{and_v(v:pk(E),pk(F)),pk(D)},{pk(C),pk(B)}})#y7cmc3xl`
Default compilation: `tr(A,{{pk(D),pk(C)},{pk(B),and_v(v:pk(E),pk(F))}})#vv7casf3`

sanket1729 commented on examples/taproot.rs:
```diff
@@ -0,0 +1,100 @@
+    assert_eq!(priv_addr, expected_addr);
+
+    // Computing the underlying script should fail for Taproot Descriptor
+    assert!(private_desc.explicit_script().is_err());
```
We should remove this

sanket1729 commented on examples/taproot.rs:
```diff
@@ -0,0 +1,100 @@
+
+        // Check if the constructed TapTree is correctly built
+        let expected_taptree: TapTree<bitcoin::XOnlyPublicKey> = TapTree::Tree(
+            Arc::new(TapTree::Leaf(Arc::new(Miniscript::<bitcoin::XOnlyPublicKey, Tap>::from_str_insane("and_v(v:pk(86fe856dc0f9adc12b863d13d03d13dec99bd1112d0425668a6f7237238aa485),older(9))").unwrap()))),
```
You should deal with abstract keys in the example and only translate them in the last step

SarcasticNastik replied:
Makes sense regarding the tutorial, thanks for the advice!

SarcasticNastik replied:
From what I understand, you are suggesting me to work with placeholder (say `String`) values instead of keys and translate them later on using `translate_pk` function. This would require me to change the accessibility of `Descriptor::translate_pk` to public.

sanket1729 reviewed (commented):
Last iteration(I think :P)

SarcasticNastik commented:
@sanket1729 I've added the document explaining how the private compilation is efficient too and refactored the example. 
I'd also appreciate comments on the doc and would love to address the ambiguities and issues, if any.

sanket1729 commented on examples/taproot.rs:
```diff
@@ -0,0 +1,145 @@
+
+    let pol: Concrete<String> = Concrete::from_str(&pol_str).unwrap();
+
+    let desc = pol.compile_tr(Some("UNSPENDABLE_KEY".to_string())).unwrap();
```
We should explain the argument here. Explain that UNSPENDABLE KEY is only when no internal key is found. Which is not the case in this example.

sanket1729 commented on src/policy/concrete.rs:
```diff
@@ -248,6 +248,9 @@ impl<Pk: MiniscriptKey> Policy<Pk> {
     /// the respective miniscripts. A Huffman Tree is created from this vector which optimizes over
     /// the probabilitity of satisfaction for the respective branch in the TapTree.
+    ///
+    /// Refer to [doc/Tr compiler.pdf] in the root of the repository to understand why such
```
Ideally, we also want the source code of this uploaded somewhere where I can copy/edit it. In case, we find some errors in the document, there would be no way to edit it.

Can you create a new repo with the code for this? And also link the source here

SarcasticNastik replied:
I've made the gist [here](https://gist.github.com/SarcasticNastik/9e70b2b43375aab3e78c51e09c288c89).

sanket1729 commented on Cargo.toml:
```diff
@@ -52,3 +53,7 @@ required-features = ["std"]
 required-features = ["std"]
+
+[[example]]
+name = "taproot"
```
You would also need to add a line for this in test.sh so that it is checked in CI

SarcasticNastik replied:
Right, thanks!

sanket1729 reviewed (commented):
Last few comments. Almost there

sanket1729 reviewed (approved):
ACK ef4249d0989dac009c99fada4c5324c523019fd4
